blog/2022-04-11-db-modules.mdx

```diff
@@ -0,0 +1,103 @@
+
+We quickly ruled out a DLL-like approach as it was too confusing in the other direction. Hard to find relevant information because *everything* has to be done with complex queries involving multiple joins, self joins, and precise where clauses.
+
+The `#include`-style was seductive because it wouldn't actually require any changes to how a "normal" software engineering team would manage their database, but we noted its error-prone nature in the case of complex relationships between various AWS services (the end user has to figure out which migration snippet to add to their migration system in what order), the lack of control on how fixes/upgrades to the schema would be applied, and the complexity of all of the snippets we would need to generate for the upgrades (each new version would need to have a migration snippet from at least the prior one, but if the prior one was a mistake you would want a way for a version bump from two prior versions, etc). It would *probably* be a quagmire long-term, is what we concluded.
+
+That leaves the three closely-related options of module tree, module list, and package manager style systems. The module tree approach was ruled out early for two big reasons: the schema name rewriting would make it hard to know "which" version of the same package is actually which, but more importantly you can create impossible data scenarios with schema replication like that. If version A of an EC2 module and version B both have a record for an EC2 instance using an `m4.large` machine, and the user edits it to make it a `t2.micro` in version A and edits it to be a `m4.4xlarge` in version B, which edit is the one to apply? What was even intended in this case? (Likely, it would be a confusion of their own thinking they had two instances instead of one running.)
+
+Even if only one was changed there could still be questions on intent, and if certain features are available in one versus another and the code backing the two versions conflicts despite the data not conflicting it could cause issues without any user input at all. A module dependency tree can be the best solution for code, but it's terrible for data.
+
+Finally, while having each module scope its tables into its own schema namespace would eliminate the possibility of collisions, [the majority of ORMs and migration systems do not support multiple schemas simultaneously](https://github.com/prisma/prisma/issues/1122), which would make adoption more difficult. With a metadata table associating table names to module/package names, the difficulties of knowing what tables belong to which is also somewhat mitigated.
+
```

Comment:
I think we should add here explicitly the option we are choosing even if it can be tell by discard?